### PR TITLE
feat: Add option for automatic timestep computation in wave solvers

### PR DESCRIPTION
**Pull request by acitrain**
This PR add the possibility of compute automatically the timestep for wave propagation solvers.

How to use it?
- You add the correct switch in your `XML` (`timestepStabilityLimit`=1) 
- You launch GEOS
- If the requested dt for the solver (`forceDt`) is higher than the CFL condition limit, substeps are used to remain below the condition
Also, an option is added to input source wavelets using `TableFunction`s, with the option `sourceWaveletTableNames`
There also a routine inside the PR to be wrapped using Pygeosx

### CODE REVIEW

File: src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEM.hpp
Comment (sframba) [inline src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEM.hpp]: Add doc for the method

File: src/coreComponents/physicsSolvers/wavePropagation/WaveSolverUtils.hpp
Comment (sframba) [inline src/coreComponents/physicsSolvers/wavePropagation/WaveSolverUtils.hpp]: I think there should be an extra MPI::sum to compute the global value over all MPI ranks. Or is it already done by the ReduceSum? Have you tried printing the value computed by each MPI rank?

File: src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEM.cpp
Comment (sframba) [inline src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEM.cpp]: use `GEOS_LOG_RANK_0` instead of `printf`

File: src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEM.cpp
Comment (sframba) [inline src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEM.cpp]: We should exit more gracefully, abruptly calling `exit` here might lead to unfinalized I/O and other undefined behavior. For example, could you use the `exitFlag` of the `EventManager`?

File: src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEMKernel.hpp
Comment (sframba) [inline src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEMKernel.hpp]: Can't you merge the previous step in this `do` loop and avoid this (sizeable) copy/paste?

File: src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEMKernel.hpp
Comment (sframba) [inline src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEMKernel.hpp]: 1.99 because you are using a 0.5% margin with respect to the CFL condition? Please add a small comment to explain

File: src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEMKernel.hpp
Comment (sframba) [inline src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEMKernel.hpp]: Can't you use the stiffness kernel directly to compute K.x instead of adding all this code? It seems like tere should be a lot of overlap with the usual computation of a forward step that the kernel already does...

File: src/coreComponents/physicsSolvers/wavePropagation/AcousticWaveEquationSEM.cpp
Comment (sframba) [inline src/coreComponents/physicsSolvers/wavePropagation/AcousticWaveEquationSEM.cpp]: same comments as above

File: src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEM.cpp
Comment (sframba) [inline src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEM.cpp]: Why are you taking the Min over all MPI ranks ? The deltaT (and the CFL condition) should be global, obtained by the Rayleigh quotient using the global fields, there shouldn't be a "local" timestep per rank, no?

**Review by sframba — COMMENTED** [review]
Thanks Aurélien! 
I have two main comments:
- The CFL factor should be global, not local to an MPI rank. Thus
  - The scalar products in `WaveSolverUtils` should include an MPI::sum to get a global value;
  - There should be no MPI::min call on the delta T, you should have a single global value
- There is a very large amount of code that essentially computes M^(-1).K.x. If I understand correctly, this should be extremely similar to what we do in a single forward time step. We should be able to refactor this and reuse the already available code, there is no need to add these many lines (and computations)

**Comment by tbeltzun** [comment]
Would there be a way to specify a zero timestep (standing for automatic computation for the timestep) in the xml file, and reuse that time step automatically in the computation, without having to modify xmls and relaunch the code ?

**Review by sframba — APPROVED** [review]

**Review by CusiniM — APPROVED** [review]

File: src/coreComponents/physicsSolvers/wavePropagation/shared/WaveSolverBase.cpp
Comment (rrsettgast) [inline src/coreComponents/physicsSolvers/wavePropagation/shared/WaveSolverBase.cpp:188]: Shouldn't this always be used unless `forceDt` is set?

**Review by rrsettgast — APPROVED** [review]

File: src/coreComponents/physicsSolvers/wavePropagation/shared/WaveSolverBase.cpp
Comment (acitrain) [reply src/coreComponents/physicsSolvers/wavePropagation/shared/WaveSolverBase.cpp:188]: In fact in this PR we are closer to the logic of the non-linear solvers: if you put a forceDt too big for the CFL condition the internal loop of the solver will do sub-steps to respect the CFL condition. 
We decided to put this feature as an option to let the user be able to tests some dt on is own without getting into that routine.

**Comment by CusiniM** [comment]
I have approved it but two things about this PR:
1. please don't put PRs  in the merge queue until they have all codeowners approvals.
2. IMO, the idea of `forceDt` is for the timestep size to be fixed. To let the solver adjust the timestep size one can just set a maxTiimestep size for the event instead of forcing it. So, I would argue that, if the `forceDt` is too large it's fine for the simulation to fail instead of substepping. I know that solver base sort of does this too but we have talked about removing it coz it is actually a lot easier to let the `EventManager` be the only object managing timestepping.

**Pull request merged**
